Make some suggested changes to style guide

benjaminbreier opened the pull request:
#### Summary
I applied some of the feedback received on the test run of the Airbnb style guide to be reviewed.

#### How I addressed each feedback

> Each line should have a maximum column width of 100 characters
> ⚠️ We're currently doing 120, but I don't mind this too much

I changed it to 120. I think its more important that we have a max, than what the max is.

> Use 2 spaces to indent lines
> 🛑 I disagree with this. Spaces are obviously more visually consistent, but we would be forcing a consistency that might be inconvenient for some folks (see https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/). We kept 4 spaces because it was already the case to avoid massive changes, but if we do change indentation then it should be to tabs

I removed this entirely.

> Use PascalCase for type and protocol names, and lowerCamelCase for everything else
> ⚠️ We already do this, but with exceptions: properties with no semantics to their names, or better their semantic is the name, should not be constrained by this (DTOs are a good example, check out the API models in our repo)

This is not automated, so I changed the wording in the README to reflect the feedback.

> Names should be written with their most general part first and their most specific part last
> ⚠️ I think that it's common practice to leave the type part of the name last. Apart from that, not opposed to it

I think based on the existing wording, this already fits just fine, so I left it unchanged.

> Event-handling functions should be named like past-tense sentences
> ⚠️ I don't necessarily agree with this (handle prefix for example gives a good hint about the nature of the method), but I can live with it

Yea, I agree, handle does work in many cases. None of the naming is automated so it can be taken as a strong suggestion :)

> Avoid *Controller in names of classes that aren't view controllers
> ⚠️ Kinda agree, but we followed the previous naming style here and didn't want to cause massive changes for it

Totally agree, modified the wording to reflect this.

> Don't include types where they can be easily inferred
> ⚠️ Agree, but with the exception of properties

I left this in. My rationale is that in the case of properties, the type is always included on the right hand side of the declaration when it can be omitted from the left, so it is never ambiguous. I hope that's alright

> Separate [long](https://github.com/airbnb/swift#column-width) function declarations with line breaks before each argument label and before the return signature

> 🛑 I agree with it apart from where to put the return signature: this style gives no clear distinction between arguments and the return signature and it's harder to spot if a function has a return or not, especially when effects (throws, async) are included. That's why we do follow this rule but put the return signature on the previous indentation level and with the closing brace on the same line

I've done my best to modify the rule appropriately based on this feedback.

> Infix operators should have a single space on either side
> ⚠️ Yes but I would apply this to range operators too

I actually disagree, but I don't mind much and went ahead and changed this to apply to range operators.

> Avoid performing any meaningful or time-intensive work in init()
> ⚠️ I generally agree, and totally agree if this means to not wait for that work in init. But I think in some cases it makes sense to trigger it (though I can be okay with enforcing this)

Yes, and this kind of style guide rule starts to creep into "best practices" recommendations. It is generally correct, but there could be exceptions. I don't think it is automated, either. We could leave it in as a guideline, or remove it entirely.



> Extract complex callback blocks into methods
> ⚠️ I agree with the general sentiment, but I'm not sure that it should be a rule. Some times it might make sense to extract part of it to well-defined logical operations instead of all of it. async/await greatly reduces the need for this anyway

It does. I agree with the thrust-- reading through someone else's code to figure out what it is doing is much more difficult when we have to leave the context to follow a dependency chain to see what a closure is doing unnecessarily. There will always be times where this is required, but I think this is another case of a "mostly true guideline".

> Access control should be at the strictest level possible
> ⚠️ Yes, but sometimes you design APIs before you use them so you can prepare the access control for that usage. If we append "for the intended usage" at the end I agree

I think that's reasonable. This is a small concern for us now, it becomes much larger when you are dealing with a very large codebase (1m+ lines of swift) and access control begins to have a meaningful impact on build times. I still believe the thrust is correct, and we could argue that your exception still follows the wording of the rule. I added "for the intended usage" to the README.

> Use Swift's automatic enum values unless they map to an external source
> 🛑 I actually agree with the sentiment, but there's a but: when we have mappings to an external source, we make a separate type explicitly for that. Having the separate type, the fact that changing it is dangerous is already explicit, so there's no need to be explicit about the rawValues

If I understand correctly, you are saying that in these situations you write two enums, one for the API's data model and one for the client's, and some glue to convert between the two, thus making the external mapping explicit. I think that's a reasonable approach to solve the same problem, and modified the wording accordingly.

> Never use the default case when switching over an enum
> ⚠️ If the rule is explicitly about a single enum, then ok (even a big one is still doable). Sometimes we do switch on tuples of enums and not having a default is impractical

Yep, once you are switching over a combination this is untenable. I modified it to be specific to single enums.

> Alphabetize and deduplicate module imports within a file. Place all imports at the top of the file below the header comments. Do not add additional line breaks between import statements. Add a single empty line before the first import and after the last import

> ⚠️ Agree with most of this, but two things:
>   Doesn't make sense to have an empty line before the first import if there's nothing else before it (and it's our case)
>   I prefer putting imports for specific things (like SwiftUI live previews) next to where it's needed, but this can be ok

I removed the guidance about the empty line (could not find an automated rule that was enforcing it). I think we'll keep preferring imports at the top of the file for now.

> Declarations that include scopes spanning multiple lines should be separated from adjacent declarations in the same scope by a newline
> ⚠️ Newlines are used to visually group sections of code, so if we have say 10 similar properties that are all single-line apart one it would feel weird to make that one break the grouping

Properties don't always include scopes spanning multiple lines, and when they do I think it is generally fine to give them some spacing. I agree that single line declarations of properties should not have spaces between them, and I believe that is not the intent of this rule.

>  Use // MARK: to separate the contents of type definitions and extensions into the sections listed below, in order

> 🛑 I agree with the spirit, but disagree with the grouping and/or ordering. What I've been pushing for is something like
```
type MyType {
  // Stored properties (split by context)
  // Lifecycle
  // `override`s, ordered by superclass hierarchy (most general superclass first)
}

// one `extension MyType` per access control of computed properties/methods
// one `extension MyType: MyProtocol` per conformance
```

This is tough. I do not mind the extension method of grouping code, although I think it serves effectively the same purpose as a pragma mark. That said, SwiftFormat provides the tooling to automate adding pragma marks and not grouping things into extensions. Would you be okay with letting it group into pragma marks as described and letting the extension division be optional or preferred?

> Within each top-level section, place content in the following order

> 🛑 Kinda agree with giving an order, but not with the specific order. I would almost do the exact opposite of the proposed one

Also tough. As far as I could see, there is no way to customize the order of these. I have considered / argued about this particular order in the past and I'd be happy to discuss / justify it if that might help assuage your concerns. Apart from that, the question becomes: would you rather have an automatically enforced order, even if it isn't the exact order you'd have chosen yourself? Or would you forego the automatically enforced order for that reason. I prefer the former.

> Add empty lines between property declarations of different kinds
> ⚠️ I don't mind too much, though I feel contextual grouping should be stronger than type of symbol

I didn't change anything related to this, I think you make a fair point, but contextual grouping isn't possible to automate in any way I can think of.

> Computed properties and properties with property observers should appear at the end of the set of declarations of the same kind

> ⚠️ I did say before that I would put computed properties separate, so that doesn't apply here. For observers, I feel like this is artificial and the order shouldn't depend on how big the symbol is in code

I think its helpful to quickly parse what amounts to the interface of a class or struct and what it is, i.e. I can see that there are 5 properties injected into it at initialization and two computed properties which are derived from X, Y, and Z almost at a glance if the order is set up like this.

Commit:
Make some suggested changes to style guide

Commit:
Further response to feedback

DeFrenZ commented:
> > Names should be written with their most general part first and their most specific part last
> > ⚠️ I think that it's common practice to leave the type part of the name last. Apart from that, not opposed to it
> 
> I think based on the existing wording, this already fits just fine, so I left it unchanged.

⚠️ The examples mention that `bodyRightMargin` is not right, but the "type" (`CGFloat`) here is "margin" really
 
> > Event-handling functions should be named like past-tense sentences
> > ⚠️ I don't necessarily agree with this (handle prefix for example gives a good hint about the nature of the method), but I can live with it
> 
> Yea, I agree, handle does work in many cases. None of the naming is automated so it can be taken as a strong suggestion :)

⚠️ Ok but suggestion towards "handleX" or "didY"? 😬 
 
> > Avoid *Controller in names of classes that aren't view controllers
> > ⚠️ Kinda agree, but we followed the previous naming style here and didn't want to cause massive changes for it
> 
> Totally agree, modified the wording to reflect this.

⚠️ If we want to "clean them up" (which I'm not sure there's an agreement on), we should choose the alternative to migrate to though

> > Don't include types where they can be easily inferred
> > ⚠️ Agree, but with the exception of properties
> 
> I left this in. My rationale is that in the case of properties, the type is always included on the right hand side of the declaration when it can be omitted from the left, so it is never ambiguous. I hope that's alright

⚠️ I understand, and that is the rationale that the rule includes as well. But mine instead is that _signatures_ and _declarations_ are more about their structure than their value. e.g.
```swift
func doThing(a: Int = 2, b: String = "c") { ... }
```
or better when multi-lined
```swift
func doThing(
    a: Int = 2,
    b: String = "c")
{ ... }
```
in here what we want to know _first_ is what type each argument is. If there is a default for it, and that from a default we can understand the type of the argument, is not that relevant. The language itself doesn't allow to do `a = 2` for that same reason.
On a type though, the language does allow it. So we can do things like
```swift
struct Thing {
    var a = 2
    var b = "c"
    var c = Foo.bar
}
```
but I find that determining the _signature_ of the properties is not as immediate (literals could be any of their `ExpressibleBy*Literal` conformers, `static` members don't necessarily return a `self`) and the default _value_ (or fixed, if `let`) of a property isn't as important as its type. So to highlight the type I'm for always being explicit about it, like
```swift
struct Thing {
    var a: Int = 2
    var b: String = "c"
    var c: Foo = .bar
}
```
there's also the case where you _need_ it in some properties (e.g. no default value) but not in others, and that asymmetry hurts the readability even more imho

all of this doesn't apply to code and not declarations though

> > Avoid performing any meaningful or time-intensive work in init()
> > ⚠️ I generally agree, and totally agree if this means to not wait for that work in init. But I think in some cases it makes sense to trigger it (though I can be okay with enforcing this)
> 
> Yes, and this kind of style guide rule starts to creep into "best practices" recommendations. It is generally correct, but there could be exceptions. I don't think it is automated, either. We could leave it in as a guideline, or remove it entirely.

✅ I would at most tweak the wording a bit to make it clear it's not an absolute rule, but something to follow generally
 
> > Extract complex callback blocks into methods
> > ⚠️ I agree with the general sentiment, but I'm not sure that it should be a rule. Some times it might make sense to extract part of it to well-defined logical operations instead of all of it. async/await greatly reduces the need for this anyway
> 
> It does. I agree with the thrust-- reading through someone else's code to figure out what it is doing is much more difficult when we have to leave the context to follow a dependency chain to see what a closure is doing unnecessarily. There will always be times where this is required, but I think this is another case of a "mostly true guideline".

✅ Same as before, maybe expanding on the examples to note where that can be not followed might help

> > Access control should be at the strictest level possible
> > ⚠️ Yes, but sometimes you design APIs before you use them so you can prepare the access control for that usage. If we append "for the intended usage" at the end I agree
> 
> I think that's reasonable. This is a small concern for us now, it becomes much larger when you are dealing with a very large codebase (1m+ lines of swift) and access control begins to have a meaningful impact on build times. I still believe the thrust is correct, and we could argue that your exception still follows the wording of the rule. I added "for the intended usage" to the README.

✅ I believe that tweak is enough to make the rule fit what I meant

> > Use Swift's automatic enum values unless they map to an external source
> > 🛑 I actually agree with the sentiment, but there's a but: when we have mappings to an external source, we make a separate type explicitly for that. Having the separate type, the fact that changing it is dangerous is already explicit, so there's no need to be explicit about the rawValues
> 
> If I understand correctly, you are saying that in these situations you write two enums, one for the API's data model and one for the client's, and some glue to convert between the two, thus making the external mapping explicit. I think that's a reasonable approach to solve the same problem, and modified the wording accordingly.

✅ Yes, your interpretation fits with what I meant. It's more about the separation of models by layer, and the intended use of _each_ model. e.g. in the API example you mentioned, a Core model _might_ have an underlying raw value that makes sense, but that should be intrinsic to the model itself and not related to its usage in other layers. If we want a mapping to `String` values to send to the API, then it makes sense to have a separate model for that (be it an `enum`, constants, or anything that fits the use-case). In this way we still avoid the unintended changes that can lead to unexpected results (if you change an API model you _know_ you're changing how you communicate with the API), but also respect the semantics of the models in each layer

> > Alphabetize and deduplicate module imports within a file. Place all imports at the top of the file below the header comments. Do not add additional line breaks between import statements. Add a single empty line before the first import and after the last import
> 
> > ⚠️ Agree with most of this, but two things:
> > Doesn't make sense to have an empty line before the first import if there's nothing else before it (and it's our case)
> > I prefer putting imports for specific things (like SwiftUI live previews) next to where it's needed, but this can be ok
> 
> I removed the guidance about the empty line (could not find an automated rule that was enforcing it). I think we'll keep preferring imports at the top of the file for now.

✅ 
 
> > Declarations that include scopes spanning multiple lines should be separated from adjacent declarations in the same scope by a newline
> > ⚠️ Newlines are used to visually group sections of code, so if we have say 10 similar properties that are all single-line apart one it would feel weird to make that one break the grouping
> 
> Properties don't always include scopes spanning multiple lines, and when they do I think it is generally fine to give them some spacing. I agree that single line declarations of properties should not have spaces between them, and I believe that is not the intent of this rule.

✅ The examples don't cover this, but my interpretation of the rule applied to properties as well. Let's go with it an we'll see if that creates confusion

> > Use // MARK: to separate the contents of type definitions and extensions into the sections listed below, in order
> 
> > 🛑 I agree with the spirit, but disagree with the grouping and/or ordering. What I've been pushing for is something like
> 
> ```
> type MyType {
>   // Stored properties (split by context)
>   // Lifecycle
>   // `override`s, ordered by superclass hierarchy (most general superclass first)
> }
> 
> // one `extension MyType` per access control of computed properties/methods
> // one `extension MyType: MyProtocol` per conformance
> ```
> 
> This is tough. I do not mind the extension method of grouping code, although I think it serves effectively the same purpose as a pragma mark. That said, SwiftFormat provides the tooling to automate adding pragma marks and not grouping things into extensions. Would you be okay with letting it group into pragma marks as described and letting the extension division be optional or preferred?

🛑 I think that `MARK`s don't give a strong visual separation as much as top-level scopes, and I care enough about the declaration scope being as limited as possible and the protocol conformance ones to stick to the protocol itself (and possibly their implementation details). I don't have much requirements over the rest

> > Within each top-level section, place content in the following order
> 
> > 🛑 Kinda agree with giving an order, but not with the specific order. I would almost do the exact opposite of the proposed one
> 
> Also tough. As far as I could see, there is no way to customize the order of these. I have considered / argued about this particular order in the past and I'd be happy to discuss / justify it if that might help assuage your concerns. Apart from that, the question becomes: would you rather have an automatically enforced order, even if it isn't the exact order you'd have chosen yourself? Or would you forego the automatically enforced order for that reason. I prefer the former.

⚠️ Automatically-enforced order has a lot of value, and I'm fine with it for _most_ of the code. But not all of it, as I greatly value visual impact of the code and which intent it provides. As long as this reordering doesn't fight the "required" bits, I'm all for it

> > Add empty lines between property declarations of different kinds
> > ⚠️ I don't mind too much, though I feel contextual grouping should be stronger than type of symbol
> 
> I didn't change anything related to this, I think you make a fair point, but contextual grouping isn't possible to automate in any way I can think of.

⚠️ The rule talks only about `static` vs instance, and I can be fine with separating those (though not sure it's worth a rule). Are there any other distinctions?

> > Computed properties and properties with property observers should appear at the end of the set of declarations of the same kind
> 
> > ⚠️ I did say before that I would put computed properties separate, so that doesn't apply here. For observers, I feel like this is artificial and the order shouldn't depend on how big the symbol is in code
> 
> I think its helpful to quickly parse what amounts to the interface of a class or struct and what it is, i.e. I can see that there are 5 properties injected into it at initialization and two computed properties which are derived from X, Y, and Z almost at a glance if the order is set up like this.

⚠️ I agree that it is helpful, but I'm not sure that this rule actually helps with it:
> I can see that there are 5 properties injected into it at initialization 

this is about the `init`, which is not involved in this rule
> and two computed properties which are derived from X, Y, and Z 

to know what the computed properties _actually_ depend on you need to read their implementation, which might be non-trivial. I would stick with moving computed properties further away as suggested before, as they're _always_ derived from something anyway and knowing exactly on what is less helpful than know what they _can_ be derived from (the stored properties)

-----

I think we're getting closer to something we can merge in 😎💪

benjaminbreier commented:
> > > Names should be written with their most general part first and their most specific part last
> > > ⚠️ I think that it's common practice to leave the type part of the name last. Apart from that, not opposed to it
> > 
> > 
> > I think based on the existing wording, this already fits just fine, so I left it unchanged.
> 
> ⚠️ The examples mention that `bodyRightMargin` is not right, but the "type" (`CGFloat`) here is "margin" really
>

**I disagree, margin is not a type, it is part of the semantic meaning of the variable. The type would be Float or Double. Also, I think those examples are meant to demonstrate the described big endian order, i.e. `bodyRightMargin` is incorrect not because it ends with `Margin`, but because `Margin` is less specific than `Right`, which is why `bodyMarginRight` is correct**

> > > Event-handling functions should be named like past-tense sentences
> > > ⚠️ I don't necessarily agree with this (handle prefix for example gives a good hint about the nature of the method), but I can live with it
> > 
> > 
> > Yea, I agree, handle does work in many cases. None of the naming is automated so it can be taken as a strong suggestion :)
> 
> ⚠️ Ok but suggestion towards "handleX" or "didY"? 😬
> 
**didY**
> > > Avoid *Controller in names of classes that aren't view controllers
> > > ⚠️ Kinda agree, but we followed the previous naming style here and didn't want to cause massive changes for it
> > 
> > 
> > Totally agree, modified the wording to reflect this.
> 
> ⚠️ If we want to "clean them up" (which I'm not sure there's an agreement on), we should choose the alternative to migrate to though
>

**I suspect this will vary widely from case to case. We could take some examples from our actual codebase to give guidelines**
> > > Don't include types where they can be easily inferred
> > > ⚠️ Agree, but with the exception of properties
> > 
> > 
> > I left this in. My rationale is that in the case of properties, the type is always included on the right hand side of the declaration when it can be omitted from the left, so it is never ambiguous. I hope that's alright
> 
> ⚠️ I understand, and that is the rationale that the rule includes as well. But mine instead is that _signatures_ and _declarations_ are more about their structure than their value. e.g.
> 
> ```swift
> func doThing(a: Int = 2, b: String = "c") { ... }
> ```
> 
> or better when multi-lined
> 
> ```swift
> func doThing(
>     a: Int = 2,
>     b: String = "c")
> { ... }
> ```
> 
> in here what we want to know _first_ is what type each argument is. If there is a default for it, and that from a default we can understand the type of the argument, is not that relevant. The language itself doesn't allow to do `a = 2` for that same reason. On a type though, the language does allow it. So we can do things like
> 
> ```swift
> struct Thing {
>     var a = 2
>     var b = "c"
>     var c = Foo.bar
> }
> ```
> 
> but I find that determining the _signature_ of the properties is not as immediate (literals could be any of their `ExpressibleBy*Literal` conformers, `static` members don't necessarily return a `self`) and the default _value_ (or fixed, if `let`) of a property isn't as important as its type. So to highlight the type I'm for always being explicit about it, like
> 
> ```swift
> struct Thing {
>     var a: Int = 2
>     var b: String = "c"
>     var c: Foo = .bar
> }
> ```
> 
> there's also the case where you _need_ it in some properties (e.g. no default value) but not in others, and that asymmetry hurts the readability even more imho
> 
> all of this doesn't apply to code and not declarations though
> 
**I think we'll have to agree to disagree here-- given the yellow emoji, are you okay with letting this one go?**
> > > Avoid performing any meaningful or time-intensive work in init()
> > > ⚠️ I generally agree, and totally agree if this means to not wait for that work in init. But I think in some cases it makes sense to trigger it (though I can be okay with enforcing this)
> > 
> > 
> > Yes, and this kind of style guide rule starts to creep into "best practices" recommendations. It is generally correct, but there could be exceptions. I don't think it is automated, either. We could leave it in as a guideline, or remove it entirely.
> 
> ✅ I would at most tweak the wording a bit to make it clear it's not an absolute rule, but something to follow generally
> 
**I think the word "Avoid" does carry that connotation to a degree, but I will try to make the wording a bit more explicit.
> > > Extract complex callback blocks into methods
> > > ⚠️ I agree with the general sentiment, but I'm not sure that it should be a rule. Some times it might make sense to extract part of it to well-defined logical operations instead of all of it. async/await greatly reduces the need for this anyway
> > 
> > 
> > It does. I agree with the thrust-- reading through someone else's code to figure out what it is doing is much more difficult when we have to leave the context to follow a dependency chain to see what a closure is doing unnecessarily. There will always be times where this is required, but I think this is another case of a "mostly true guideline".
> 
> ✅ Same as before, maybe expanding on the examples to note where that can be not followed might help
> 
**Fair, feel free to provide an example or two**
> > > Access control should be at the strictest level possible
> > > ⚠️ Yes, but sometimes you design APIs before you use them so you can prepare the access control for that usage. If we append "for the intended usage" at the end I agree
> > 
> > 
> > I think that's reasonable. This is a small concern for us now, it becomes much larger when you are dealing with a very large codebase (1m+ lines of swift) and access control begins to have a meaningful impact on build times. I still believe the thrust is correct, and we could argue that your exception still follows the wording of the rule. I added "for the intended usage" to the README.
> 
> ✅ I believe that tweak is enough to make the rule fit what I meant
> 
> > > Use Swift's automatic enum values unless they map to an external source
> > > 🛑 I actually agree with the sentiment, but there's a but: when we have mappings to an external source, we make a separate type explicitly for that. Having the separate type, the fact that changing it is dangerous is already explicit, so there's no need to be explicit about the rawValues
> > 
> > 
> > If I understand correctly, you are saying that in these situations you write two enums, one for the API's data model and one for the client's, and some glue to convert between the two, thus making the external mapping explicit. I think that's a reasonable approach to solve the same problem, and modified the wording accordingly.
> 
> ✅ Yes, your interpretation fits with what I meant. It's more about the separation of models by layer, and the intended use of _each_ model. e.g. in the API example you mentioned, a Core model _might_ have an underlying raw value that makes sense, but that should be intrinsic to the model itself and not related to its usage in other layers. If we want a mapping to `String` values to send to the API, then it makes sense to have a separate model for that (be it an `enum`, constants, or anything that fits the use-case). In this way we still avoid the unintended changes that can lead to unexpected results (if you change an API model you _know_ you're changing how you communicate with the API), but also respect the semantics of the models in each layer
> 
**I wonder if we should add a naming guidelines to include `API` in the name of all `API models` to make this even more clear?**
> > > Alphabetize and deduplicate module imports within a file. Place all imports at the top of the file below the header comments. Do not add additional line breaks between import statements. Add a single empty line before the first import and after the last import
> > 
> > 
> > > ⚠️ Agree with most of this, but two things:
> > > Doesn't make sense to have an empty line before the first import if there's nothing else before it (and it's our case)
> > > I prefer putting imports for specific things (like SwiftUI live previews) next to where it's needed, but this can be ok
> > 
> > 
> > I removed the guidance about the empty line (could not find an automated rule that was enforcing it). I think we'll keep preferring imports at the top of the file for now.
> 
> ✅
> 
> > > Declarations that include scopes spanning multiple lines should be separated from adjacent declarations in the same scope by a newline
> > > ⚠️ Newlines are used to visually group sections of code, so if we have say 10 similar properties that are all single-line apart one it would feel weird to make that one break the grouping
> > 
> > 
> > Properties don't always include scopes spanning multiple lines, and when they do I think it is generally fine to give them some spacing. I agree that single line declarations of properties should not have spaces between them, and I believe that is not the intent of this rule.
> 
> ✅ The examples don't cover this, but my interpretation of the rule applied to properties as well. Let's go with it an we'll see if that creates confusion
> 
> > > Use // MARK: to separate the contents of type definitions and extensions into the sections listed below, in order
> > 
> > 
> > > 🛑 I agree with the spirit, but disagree with the grouping and/or ordering. What I've been pushing for is something like
> > 
> > 
> > ```
> > type MyType {
> >   // Stored properties (split by context)
> >   // Lifecycle
> >   // `override`s, ordered by superclass hierarchy (most general superclass first)
> > }
> > 
> > // one `extension MyType` per access control of computed properties/methods
> > // one `extension MyType: MyProtocol` per conformance
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is tough. I do not mind the extension method of grouping code, although I think it serves effectively the same purpose as a pragma mark. That said, SwiftFormat provides the tooling to automate adding pragma marks and not grouping things into extensions. Would you be okay with letting it group into pragma marks as described and letting the extension division be optional or preferred?
> 
> 🛑 I think that `MARK`s don't give a strong visual separation as much as top-level scopes, and I care enough about the declaration scope being as limited as possible and the protocol conformance ones to stick to the protocol itself (and possibly their implementation details). I don't have much requirements over the rest
> 

**I did remove this rule entirely. Let's continue discussion of rules that are not automatable offline.**

> > > Within each top-level section, place content in the following order
> > 
> > 
> > > 🛑 Kinda agree with giving an order, but not with the specific order. I would almost do the exact opposite of the proposed one
> > 
> > 
> > Also tough. As far as I could see, there is no way to customize the order of these. I have considered / argued about this particular order in the past and I'd be happy to discuss / justify it if that might help assuage your concerns. Apart from that, the question becomes: would you rather have an automatically enforced order, even if it isn't the exact order you'd have chosen yourself? Or would you forego the automatically enforced order for that reason. I prefer the former.
> 
> ⚠️ Automatically-enforced order has a lot of value, and I'm fine with it for _most_ of the code. But not all of it, as I greatly value visual impact of the code and which intent it provides. As long as this reordering doesn't fight the "required" bits, I'm all for it
> 
**Yea, let's keep workshopping this rule and the one above until we find something we can agree on AND automate. What I really hope to avoid is placing a burden on you and other engineers in code reviews to be enforcing style. For now we will leave this out entirely.**
> > > Add empty lines between property declarations of different kinds
> > > ⚠️ I don't mind too much, though I feel contextual grouping should be stronger than type of symbol
> > 
> > 
> > I didn't change anything related to this, I think you make a fair point, but contextual grouping isn't possible to automate in any way I can think of.
> 
> ⚠️ The rule talks only about `static` vs instance, and I can be fine with separating those (though not sure it's worth a rule). Are there any other distinctions?
> 
**I interpret it as grouping typealiases together, grouping imports together, grouping variables together, and so on. There is an emergent, cascading set of ordering rules encoded into these rules that isn't explicitly defined, maybe it would be helpful to define it more explicitly at some point? But even as I write that, my goal is to make that automated, so that engineers don't need to consider it while writing code, so maybe that is an anti-pattern with that north star in mind.**

> > > Computed properties and properties with property observers should appear at the end of the set of declarations of the same kind
> > 
> > 
> > > ⚠️ I did say before that I would put computed properties separate, so that doesn't apply here. For observers, I feel like this is artificial and the order shouldn't depend on how big the symbol is in code
> > 
> > 
> > I think its helpful to quickly parse what amounts to the interface of a class or struct and what it is, i.e. I can see that there are 5 properties injected into it at initialization and two computed properties which are derived from X, Y, and Z almost at a glance if the order is set up like this.
> 
> ⚠️ I agree that it is helpful, but I'm not sure that this rule actually helps with it:
> 
> > I can see that there are 5 properties injected into it at initialization
> 
> this is about the `init`, which is not involved in this rule
> 
> > and two computed properties which are derived from X, Y, and Z
> 
> to know what the computed properties _actually_ depend on you need to read their implementation, which might be non-trivial. I would stick with moving computed properties further away as suggested before, as they're _always_ derived from something anyway and knowing exactly on what is less helpful than know what they _can_ be derived from (the stored properties)
> 
> I think we're getting closer to something we can merge in 😎💪

**I think my response was about the ordering section in total rather than that specific rule in a vacuum, but I think your point is reasonable. I do have a preference for keeping the properties together, with computed towards the end, because it does keep the definitions of what the computed properties are derived from nearby. But its a small thing, far from the most important rule on this list.**

Given that I have already removed both of the automated rules you had 🛑 s on, if I modify a couple of the rules' wording according to my responses above, are you okay with moving to the next step in this process (adding the relevant files to the app repo, applying the changes, and creating a PR)?

Commit:
Couple more wording changes to make things clearer

DeFrenZ commented:
> I disagree, margin is not a type, it is part of the semantic meaning of the variable. The type would be Float or Double. Also, I think those examples are meant to demonstrate the described big endian order, i.e. bodyRightMargin is incorrect not because it ends with Margin, but because Margin is less specific than Right, which is why bodyMarginRight is correct

I think that "margin" is part of the semantics of the variable _only_ because a specific-enough type wasn't used/available for this. But not that big of a deal

> didY

I would've liked the other one better, but not going to stop for this (it's still sensible enough)

> I suspect this will vary widely from case to case. We could take some examples from our actual codebase to give guidelines

Fair. But if we include this it means we _want_ to gradually clean them up, and should create a vocabulary of what each noun represents in kinds of type of behaviour

> I think we'll have to agree to disagree here-- given the yellow emoji, are you okay with letting this one go?

Actually, for this one I will upgrade it to a 🛑 as it's something I believe affects the readability of types' storage noticeably. Though as every thing that I'm opposed to, if the rest of the team agrees I will have to accept it regardless (I have strong opinions but I'm not a dictator)

> I wonder if we should add a naming guidelines to include API in the name of all API models to make this even more clear?

We do respect this at the moment, but I'm not even sure it's something we _should_ indeed do. But if we don't, then we probably want to use the namespace consistently (e.g. `IRLAPI.GroupV3`)? Don't have a strong opinion here (yet)

> I did remove this rule entirely. Let's continue discussion of rules that are not automatable offline.

Sure, or anyway not something that stops this

> Yea, let's keep workshopping this rule and the one above until we find something we can agree on AND automate. What I really hope to avoid is placing a burden on you and other engineers in code reviews to be enforcing style. For now we will leave this out entirely.

This is definitely something I'd love as well. Less friction everywhere and faster feedback cycles are something great 🙌 

> I interpret it as grouping typealiases together, grouping imports together, grouping variables together, and so on. There is an emergent, cascading set of ordering rules encoded into these rules that isn't explicitly defined, maybe it would be helpful to define it more explicitly at some point? But even as I write that, my goal is to make that automated, so that engineers don't need to consider it while writing code, so maybe that is an anti-pattern with that north star in mind.

Defined like this, I'm okay with it

-----

> Given that I have already removed both of the automated rules you had 🛑 s on, if I modify a couple of the rules' wording according to my responses above, are you okay with moving to the next step in this process (adding the relevant files to the app repo, applying the changes, and creating a PR)?

There's only one point I feel should change, but apart from that yes

Commit:
Only prefer inferred types in local scopes (explicit in properties)

benjaminbreier commented:
Okay @DeFrenZ, I've made the change so that inferred types are only preferred in local scopes.
